### ISSUE TITLE
Recursion versus iteration

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#8873]
Recursion versus iteration<br>
<br>
https://ift.tt/vOXilKA<br>
<br>
<br>
<br>
<div>
<div>
<div>
<div>
<div>
<p>The first programming language to allow functions to be recursively defined was McCarthy&rsquo;s LISP in 1959. Its introduction was not controversial: nobody but John McCarthy had any say in what the language was going to be. In addition to his work on LISP, McCarthy was on the committee finalizing the Algol language in 1959 and 1960. In spite of the fact that a majority was opposed to it, the definition of Algol 60 ended up allowing recursively defined procedures. In [<a href="https://vanemden.wordpress.com/2014/07/30/recursion-versus-iteration/#wordpress">1</a>] I gave an acount of how this happened.</p>
<p>Why was recursion such a big deal? For us this is hard to understand: for decades the programming language C, not exactly a paradigm of avant-garde, has allowed recursion. Still, remnants of unease remain. Still, in some introductory courses, recursion plays the role of a <em>pons asinorum</em>. Have instructors been traumatized by a sarcastic teacher finding, oh horrors, a circular definition in their high-school essay? In <em>this</em> essay I&rsquo;m going to explore recursion by contrasting it to its opposite, iteration.</p>
<p>I propose that the distinction between recursion and iteration for the same computational task is related to the possibility noted by Robert Kowalski that the same induction rule can be used with bottom-up and top-down reasoning and that the choice between these possibilities has algorithmic consequences [<a href="https://vanemden.wordpress.com/2014/07/30/recursion-versus-iteration/#AeqLplusC">2</a>].</p>
<p>For example, here is an inductive characterization of the Fibonacci numbers:</p>
<blockquote>
<p>fib(0) = 0 and fib(1) = 1 and<br>
for all a,b,n:<br>
fib(n) = a and fib(n+1) = b -&gt; fib(n+2) = a+b</p>
</blockquote>
<p>which says that whenever two successive Fibonacci numbers are known, the next can be found by adding them. This reading of the induction rule is the use of the rule in bottom-up reasoning mode. It suggests an iterative program. In C it could be:</p>
<pre>int fib(int n) {
  if (n == 0 || n == 1) return n;
  { int a = 0, b = 1, i = 1;
    while (i++ &lt; n) {
      int temp = a; a = b; b += temp;
      // b is the i-th Fibonacci number
    }
    // b is the n-th Fibonacci number
    return b;
  }
}
</pre>
<p>The induction rule can be put in a different format by having the implication arrow point the other way. This suggests its use in top-down reasoning mode:</p>
<blockquote>
<p>fib(0) = 0 and fib(1) = 1 and<br>
for all a,b,n:<br>
fib(n) = a+b &lt;- fib(n-1) = b and fib(n-2) = a</p>
</blockquote>
<p>This format of the rule transcribes to a recursively defined function:</p>
<pre>int fib(int n) {
  if (n == 0 || n == 1) return n;
  return fib(n-1) + fib(n-2);
}
</pre>
<p>More succinct, more obviously seen to be correct with respect to the induction rule, and egregiously inefficient.</p>
<p>Clearly, a strike against recursion.</p>
<p>That a majority of the Algol committee was opposed to allowing recursion in the language was probably not because of Fibonacci numbers. The justification was that it was not clear how to implement it. Moreover, Algol was meant to be for numerical computation and there seemed to be no need for recursion in that type of application. In the fullness of time everyone was convinced of the virtues of recursion by Hoare&rsquo;s Quicksort [<a href="https://vanemden.wordpress.com/2014/07/30/recursion-versus-iteration/#qsort">3</a>].</p>
<p>That majority position against recursion is interesting because it is wrong on two counts.</p>
<ol>
<li>Adaptive Simpson integration is numerical and every bit as recursive as Quicksort is. This shows that even numerical analysts have a limited view of numerical analysis.</li>
<li>How convincingly recursive is Quicksort anyway? More on this point below.</li>
</ol>
<p>For a long time I thought that iterative Quicksort, while possible, would be too horrible to contemplate. Until I took the trouble to write it.</p>
<p>First, as a basis for comparison, the convincingly and elegantly recursive Quicksort.</p>
<pre>void qSort1(int a[], int p, int q) {
// sorts a[p..q-1]
  if (p &gt;= q-1) // one element or fewer; do nothing
    return;
  int m = partition(a, p, q); // a[m] is in place
  qSort1(a, p, m); // sort a[p..m-1]
  qSort1(a, m+1, q); // sort a[m+1..q-1]
}
void qSort(int a[], int n) { // sorts a[0..n-1]
  qSort1(a, 0, n);
}
</pre>
<p>But this is only acceptable if it is acceptable that an unfavourable input results in overflow of the system stack. If not, then some modifications are called for.</p>
<p>First, the second of the two recursive calls needs to be eliminated. This is the well-known tail-recursion elimination: the observation that under certain conditions a recursive call at the end of the function body can be replaced by a transfer of control to the beginning.</p>
<p>Once this is done, we can choose which of the two segments resulting from the call to &ldquo;partition&rdquo; is to be sorted by the remaining recursive call. By choosing the smaller of the two segments we ensure that the number of pending activations of &ldquo;qsort1&rdquo; is bounded by log n for segments of length n.</p>
<p>After these modifications the elegance of the original Quicksort is somewhat tarnished:</p>
<pre>void qSort1(int a[], int p, int q) {
// sorts a[p..q-1]
  int m;
  while (p &lt; q-1) {
    m = partition(a, p, q); // a[m] is in place
    if (m-p &lt; q-(m+1)) {
      qSort1(a, p, m); p = m+1;
    }
    else {
      qSort1(a, m+1, q); q = m;
    }
  }
}
void qSort(int a[], int n) { // sorts a[0..n-1]
  qSort1(a, 0, n);
}
</pre>
<p>But, of course, still elegant.</p>
<p>Without recursion the overflow-proof version becomes something like the code shown below. It is based on a pair of stacks<br>
int lb[50] and int rb[50]<br>
for the left and right bounds of array segments that still need to be sorted. Thus typically, with k items on the stacks, we still need to sort<br>
a[lb[i] .. rb[i]-1]<br>
for i = 0, &hellip;, k-1.</p>
<pre>void qSort1(int a[], int p, int q) {
const int max = 50; int lb[max], rb[max], n = 0;
L: // to be sorted: a[p..q-1] and the n items on the stack
  if (p &lt; q-1) { // at least two elements in a[p..q-1]
    int m = partition(a, p, q);
    if (m-p &lt; q-(m+1)) { //push smaller (left) half
      lb[n] = p; rb[n] = m; n++;
      // sort right half
      p = m+1; goto L;
    } else { // push smaller (right) half
      lb[n] = m+1; rb[n] = q; n++;
      // sort left half
      q = m; goto L;
    }
  } // p &gt;= q-1, so a[p..q-1] already sorted
  if (n == 0) return; // because nothing on the stack
  n--; p = lb[n]; q = rb[n];
  goto L;
}
void qSort(int a[], int n) { // sorts a[0..n-1]
  qSort1(a, 0, n);
}
</pre>
<p>Not as concise as the recursive version, but hardly the horrible mess I had hoped for as a fan of recursion.</p>
<p>A word about that label and my use of the g-word. I found that using any of the standard control patterns resulted in superfluously stacking and unstacking or in superfluous testing. Maybe I didn&rsquo;t try hard enough. But why should I have tried at all? Keeping in mind what needs to be true at the label (see the comment there), the code writes itself.</p>
<p>For Quicksort recursion is nice to have, but hardly essential. Once you know the idea, the iterative version is just as understandable as the recursive one. But <em>to get the idea</em>, recursion is essential. Or so I thought, until a knowledgeable friend told me that Sir Anthony Hoare, the inventor of Quicksort, wrote the code for it <em>without even knowing</em> that recursive subroutines are possible [<a href="https://vanemden.wordpress.com/2014/07/30/recursion-versus-iteration/#hoare">4</a>].</p>
<p>Now that recursion is seen as not essential to Quicksort, one may wonder whether it ever is. I give two examples of programming situations where it makes more than a superficial difference.</p>
<p>Let us consider the problem of producing the decimal numeral representing a given nonnegative integer. To avoid the clutter associated with manipulating strings or lists, we produce the numeral by printing its digits.</p>
<pre>void num1(unsigned i) {
  while (i&gt;0) { print(i%10); i /= 10; }
}
void num(unsigned i) { num1(i); nl(); }
</pre>
<p>Here &ldquo;print&rdquo; prints its argument as a decimal digit; &ldquo;nl&rdquo; prints a newline. Basically the same algorithm is used in printing decimal numerals in most libraries. This algorithm prints the digits backwards, which can, and is, of course easily corrected by reversing the string produced.</p>
<p>Just out of curiosity, what does it take to print the digits immediately in the right order? Here is one way to do it. Regard the loop as an optimized recursion. The unoptimized version of the above is:</p>
<pre>void num1(unsigned i) {
  if (i&gt;0) { print(i%10); num1(i/10); }
}
void num(unsigned i) { num1(i); nl(); }
</pre>
<p>This makes it obvious that the last digit gets printed first. It also makes it obvious what to do about it: print the last digit <em>after</em> the others, as in the code below.</p>
<pre>void num1(unsigned i) {
  if (i&gt;0) { num1(i/10); print(i%10); }
}
void num(unsigned i) { num1(i); nl(); }
</pre>
<p>Two observations.</p>
<ol>
<li>To get the digits out in the correct order one would like to have temporary storage for those digits.</li>
<li>In this program by the time the last (in the correct order) digit is computed, there are as many activations of the function num1 as there are digits and none of these has completed. Each of these activations stores its own copy of the parameter that allows another digit to be computed. The allocation and deallocation happens without any code explicitly forcing this to happen. It happens at the computational cost of a function call and exit.</li>
</ol>
<p>Allocation and deallocation of storage in local copies of parameters or in local variables or arrays has advantages over the use of storage in the heap. Allocation in the heap requires finding an unoccupied segment of sufficient length. When a program has caused a considerable number of allocations and deallocations of heap storage, the available storage on the heap has become fragmented. That is, even though the total unallocated space may be plentiful, it may be time-consuming, or even impossible, to find a free segment of the required length.</p>
<p>The technique exemplified here, that of using multiple activations of the same subroutine as temporary storage, is applicable whenever the storage can be allocated and de-allocated on a last-in first-out basis. I suspect that the heap is often (mis)used without considering alternatives.</p>
<p>Here is an example. Suppose that in C we want to copy standard input to an area in the heap. As the length of standard input remains unknown until the last byte has been read, temporary storage is needed to ensure that exactly the right amount of heap storage is allocated. As the needed temporary storage can be allocated and de-allocated last-in, first-out, an array local to a recursively defined function can be harnessed to this purpose.</p>
<p>As I am running out of time and space, I will omit further explanation, while realizing that, due to the unfamiliar technique, the code may present a bit of a puzzle.</p>
<pre>#include &lt;stdio.h&gt;
#include &lt;stdlib.h&gt;

int stdin2heap1(char** p, int n, int size) {
// returns number of characters in remainder of stdin
// n characters have been read by previous invocations
// and are stored in block[] of previous invocations
  char block[size]; int c;
  int m; // for number of characters to be read by
         // future invocations
  int i = 0;
A: // i characters have been read by this invocation
   // i &lt; size
  c = getchar();
  if (c == EOF) { m = 0; *p = malloc(n + i); goto B; }
  block[i++] = c;
  if (i &lt; size) // there is space in block[]
    goto A;
  m = stdin2heap1(p, n + size, 2*size);
B: m += i;
  // m characters have been read by previous invocations
  // and by the current invocation
  while (i--) *(*p + n + i) = block[i];
  return m;
}
int stdin2heap(char** p) {
// Copy the contents of stdin to the heap.
// Return the number of characters read.
  return stdin2heap1(p, 0, 1);
}
</pre>
<p>It is essential here to acknowledge the crucial contributions of Gauthier van den Hove. At the same time the reader should be warned that any reasons for disliking it are likely to be due to my extensive modifications.</p>
<p>Too bad that the operating systems people think that one should not do this kind of thing. That is at least my conclusion from the fact that even in systems generously endowed with RAM the systems stack gets a measly 8 MB.</p>
<p>We noted Kowalski&rsquo;s use of &ldquo;bottom-up&rdquo; versus &ldquo;top-down&rdquo; to characterize forward versus backward reasoning with inductive rules. This terminology derives from parsing, as we can see in Kowalski&rsquo;s account [<a href="https://vanemden.wordpress.com/2014/07/30/recursion-versus-iteration/#histLog">5</a>] of his collaboration with Alain Colmerauer concerning logic representation of context-free grammars:</p>
<blockquote>
<p><em>We saw that different kinds of resolution applied to the resulting grammars give rise to different kinds of parsers. For example, forward reasoning with hyper-resolution performs bottom-up parsing, while backward reasoning with SL-resolution performs top-down parsing.</em></p>
</blockquote>
<p>Indeed, production rules of context-free grammars, typically recursive, read like inductive definitions.</p>
<p>Independently of this, compiler writers have identified top-down with recursive by attaching the term &ldquo;recursive-descent&rdquo; to top-down parsers. To me it seems that the natural way to define the syntax of a high-level programming language is by means of a context-free grammar and that the natural way of parsing sentences of a context-free language is with a recursive-descent parser. Therefore I am puzzled that this view is not shared by the majority of those who know more about compiler implementation than I do.</p>
<p>I was gratified to find an account of his parsing adventures by Bjarne Stroustrup , the creator of C++ [<a href="https://vanemden.wordpress.com/2014/07/30/recursion-versus-iteration/#cpp">6</a>, pp. 68&mdash;69]. In 1982, when he was planning Cfront, the precursor of C++, he wanted to use a recursive-descent parser because he liked the ease with which such parsers could be made to provide helpful error messages and because he liked to have the full power of a high-level programming language available for implementing the decisions to be made in the parser.</p>
<p>However, Stroustrup took advice from his colleagues and used instead YACC to generate a parser. Given that these colleagues were Al Aho and Steve Johnson, that advice was somewhat biased. Stroustrup continues: [<a href="https://vanemden.wordpress.com/2014/07/30/recursion-versus-iteration/#cpp">6</a>, p. 69]</p>
<blockquote>
<p><em>In retrospect, for me and C++ it was a bad mistake.</em></p>
</blockquote>
<p>A litany of problems follows and concludes with the remark that several modern C++ compilers do use recursive descent. &lt;!&ndash;</p>
<p>In this article I noted several instances of a certain bias against the top-down, recursive approach that seems to run like a muddy thread through the length of the short life of programming. &ndash;&gt;</p>
<h4>Acknowledgments</h4>
<p>Thanks to Paul McJones and Gauthier van den Hove for their careful reading and suggestions for improvement.</p>
<h4>References</h4>
<p>[<a name="wordpress" id="wordpress"></a>1] &ldquo;<a href="http://tinyurl.com/muukmht">How Recursion Got Into Programming</a>&rdquo;<br>
[<a name="AeqLplusC" id="AeqLplusC"></a>2] &ldquo;Algorithm = Logic+ Control&rdquo; by R.A. Kowalski. <em>Comm. ACM</em>, 22(7), 424&mdash;436, 1979.<br>
[<a name="qsort" id="qsort"></a>3] &ldquo;Quicksort&rdquo; by C. A. R. Hoare. <em>Comput. J.</em> 5 (1962), pp. 10&mdash;16.<br>
&ldquo;Algorithm 63: partition&rdquo; and &ldquo;Algorithm 64: Quicksort&rdquo; by C.A.R. Hoare. <em>Comm. ACM</em> 7 (1961), pp. 321&mdash;322.<br>
[<a name="hoare" id="hoare"></a>4] &ldquo;The Emperor&rsquo;s Old Clothes&rdquo; by C. A. R. Hoare. <em>Comm. ACM</em> 24 (1981), pp. 75&mdash;83.<br>
[<a name="histLog" id="histLog"></a>5] &ldquo;Logic Programming&rdquo; by R.A. Kowalski, in <em>Computational Logic</em> Dov Gabbay, J&ouml;rg Siekmann, and John Woods (eds.), volume 9 of <em>Handbook of the History of Logic</em> North-Holland, to appear.<br>
[<a name="cpp" id="cpp"></a>6] <em>The Design and Evolution of C++</em> by Bjarne Stroustrup. Addison-Wesley, 1994.</p>
<p><small>This entry was posted on July 30, 2014 at 5:24 am and is filed under <a rel="category tag" href="https://vanemden.wordpress.com/category/programming/">Programming</a>. You can follow any responses to this entry through the <a href="https://vanemden.wordpress.com/2014/07/30/recursion-versus-iteration/feed/">RSS 2.0</a> feed. You can <a href="https://vanemden.wordpress.com/2014/07/30/recursion-versus-iteration/#respond">leave a response</a>, or <a rel="trackback" href="https://vanemden.wordpress.com/2014/07/30/recursion-versus-iteration/trackback/">trackback</a> from your own site.</small></p>
</div>
</div>
</div>
</div>
</div>
<br>
<br>
<br>
<br>
<br>
<br>
via A Programmers Place<br>
<br>
October 8, 2022 at 05:02PM
